This repository contains the code used for analysis and figures for the manuscript titled: **Differentiating Mpox Infection and Vaccination Using a Validated Multiplex Orthopoxvirus IgG Serology
Assay**

Running bash scripts in main or validation folders will run all make files for generating tables and figures. Output will appear in the "fig_ouput" folders.

# **Folder: main**

[**Files:**]{.underline}

**2025-Oct-15_compiled_data.csv** - all MSD testing and MVA neutralization results

**make\_\*** - Rscripts for generating figures and tables

**MPXV_paper_functions.r** - contains functions shared between scripts in main and validation

**run_scripts.sh** - bash script to run all make\_\* scripts

[**Folders:**]{.underline}

**fig_output** - output files from running make\_\* scripts

**similarity files -** contains:

-   **MSD_antigen_domain.xlsx** - position information for ectodomain, endodomain, and cytoplasmic tail for each MSD antigen

-   **MSD antigen pairwise folder** - contain pairwise alignments between MPXV and VACV orthologs (fasta)

-   **virion image folder** - contains images for Fig 1B

# **Folder: validation (supplementary material)**

[**Files:**]{.underline}

**2025-Oct-15_MPXV_paper_validation_data** - MSD validation dataset

**2025_Aug_26_all_QC** - QC results from runs 2024-07-30 to 2025-01-31

**2025_Aug_26_all_CAL_Control_data** - Calibrator and control (1,2,3) data from runs 2024-07-30 to 2025-01-31

**make\_\*** - Rscripts for generating figures and tables in supplementary material

**run_scripts.sh** - bash script to run all make\_\* scripts

[**Folders:**]{.underline}

**assay_limits** - running scripts will generate assay_para_validated.csv, which contains the final LoD, LLoQ, and ULoQ. File is generated by running make_Fig_S1_S3_Table_S1_S2.r. Subsequent scripts
are dependent on this file.

**expected_value_table** - expected MSD results for VIG and M32

**fig_output** - output files from running make\_\* scripts
